### PR TITLE
Centralise GD client functionality.

### LINKED ISSUE #68: Add proxy support for calls to official servers.
Geometry Dash uses a cloudflare filtering system agains server IPs. This blocks most cloud IPs from accessing the server. We call the servers for:
- Song data
- Resource CDN URL.

We should have a configurable proxy that allows us to use something like https://github.com/RealistikGDPS/gd-traffic-proxy on the server side.

### PR DESCRIPTION
**Pull request by RealistikDash**
Creates a single adapter thats been abstracted away for easy GD server interaction. Also allows more customisation, allowing for features such as proxying.

### CODE REVIEW

**Comment by RealistikDash** [comment]
Tested: works fully!

File: rgdps/services/boomlings.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+GEOMETRY_DASH_URL = "https://www.boomlings.com/database"
+"""The default URL to access Geometry Dash's servers."""
+
+# This is meant to be an empty string.
+GEOMETRY_DASH_HEADER = ""
+"""The User-Agent header used by the client. This is specifically checked
+for by the firewall."""
+
+_CLOUDFLARE_ERROR_RESPONSE = "error code: 1005"
+"""A CloudFlare firewall response if your response was blocked with an empty
```
Comment (7ez) [inline rgdps/services/boomlings.py]: Cloudflare 👍

File: rgdps/services/boomlings.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+_ROBTOP_ERROR_RESPONSE = "no"
+"""Unknown fully but generally if rob does something personally againts you/your IP,
+this is what is returned."""
+
+_GENERIC_ERROR_RESPONSE = "-1"
+"""General error response code. Could be as simple as "not found"."""
+
+
+_SONG_ENDPOIT = "/getGJSongInfo.php"
```
Comment (7ez) [inline rgdps/services/boomlings.py]: ENDPOINT

File: rgdps/services/boomlings.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+# This is meant to be an empty string.
+GEOMETRY_DASH_HEADER = ""
+"""The User-Agent header used by the client. This is specifically checked
+for by the firewall."""
+
+_CLOUDFLARE_ERROR_RESPONSE = "error code: 1005"
+"""A CloudFlare firewall response if your response was blocked with an empty
+User-Agent header."""
+
+_ROBTOP_ERROR_RESPONSE = "no"
```
Comment (7ez) [inline rgdps/services/boomlings.py:54]: against

File: rgdps/services/boomlings.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        
+        return content
+    
+
+    async def get_song(self, song_id: int) -> GDStatus[IntKeyResponse]:
+        """Queries the official servers for a song with a given id. Parses
+        the response into a dictionary."""
+
+        song_info = await self.__make_post_request(
+            _SONG_ENDPOIT,
```
Comment (7ez) [inline rgdps/services/boomlings.py]: ENDPOINT

**Review by 7ez — CHANGES_REQUESTED** [review]

File: rgdps/services/boomlings.py (hunk 1 above)
Comment (7ez) [reply rgdps/services/boomlings.py]: image
smh

**Review by 7ez — APPROVED** [review]
It appears scrumptious to I

**Pull request merged**